CESQL v1 review changes

Cali0707 opened the pull request:
Addresses the comments on #1282 

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

Look at the discussion on #1282

jskeet commented on cesql/spec.md:
```diff
@@ -200,6 +200,8 @@ _Boolean_.
 When addressing an attribute not included in the input event, the subexpression referencing the missing attribute MUST evaluate to `false`.
 For example, `true AND (missingAttribute = "")` would evaluate to `false` as the subexpression `missingAttribute = ""` would be false.
 
+When addressing an attribute represented by the _Integer_ type, if the value of that attribute does not fit into a 32 bit signed integer, an error MUST be returned.
```
A CloudEvent Integer attribute can't have such a value though... from the CE Spec:

> A whole number in the range -2,147,483,648 to +2,147,483,647 inclusive. This is the range of a signed, 32-bit, twos-complement encoding. Event formats do not have to use this encoding, but they MUST only use Integer values in this range.

To put it another way: if we've got this for integers, do we need something similar for attributes represented by the _Boolean_ type but with a value which isn't true or false?

Cali0707 replied:
Ah, if it isn't possible for this value to be in an event then I think we can remove this - I hadn't been aware that that was the case

jskeet replied:
Yes, it's the CE type system rather than about specific attributes.

duglin commented on cesql/spec.md:
```diff
@@ -56,7 +56,10 @@ producer, or in an intermediary, and it can be implemented using any technology
+transparency][referential-transparency-wiki]. The primary output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_.
+The secondary output of a CESQL expression evaluation is a set of errors which occurred during evaluation. This set MAY be empty, indicating that no
+error occurred during execution of the expression. The value used by CESQL engines to represent an empty set of errors is out of the scope of this
+specification.
```
I like this better. Do you think we need to add something like:
> Generation of an error does not halt processing. This specification does not mandate how these errors are exposed to a user.

? Is the first sentence true?

Cali0707 replied:
I'm not 100% sure here, if you look at section 4.1 it states that:

> When evaluating an expression, the evaluator can operate in two _modes_, in relation to error handling:
> * Fail fast mode: When an error is triggered, the evaluation is interrupted and returns the error, without any result.
> * Complete evaluation mode: When an error is triggered, the evaluation is continued, and the evaluation of the expression returns both the result and the error(s).

But, as far as I can tell the SDKs implement the fail fast mode instead of the complete evaluation mode.

duglin replied:
w/o any result... is that true even in the SDKs and Knative? I was assuming the was result 'false'.

I'm having a memory of us discussing what happens when a nested expression errors when that expression is part of an AND - doesn't it result in a false for the nested expression and then false again due to that side of the AND being false? Or am I thinking about some other situation?

Cali0707 replied:
In terms of w/o any result, I think the problem (from a spec standpoint at least) is what type should the return value have? For example, if my query is `LEFT(missingattribute, 10)`, then should I get `(false, missing attribute error)` as my return value? Does it make more sense to return `("", missing attribute error)`?

But, similarly if my expression is `LEFT(missingattribute, 10) LIKE prefix%`, then having a return value of `(false, missing attribute error)` would be what I would expect while a return value of `("", missing attribute error)` would not be what I expect.

Perhaps it makes sense to say something along the lines of "Fail fast mode: When an error is triggered, the evaluation is interrupted and returns the zero value for the return type of the root operation, along with the error". WDYT @duglin ?

duglin commented on cesql/spec.md:
```diff
@@ -70,7 +73,8 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
+When used as a filter predicate, the expression output value MUST be a _Boolean_. If the output value is not a _Boolean_ or any errors are returned,
+the event MUST NOT pass the filter.
 
 <!-- TODO -->
```
do we still need this TODO ?

Cali0707 replied:
Removed

duglin commented on cesql/spec.md:
```diff
@@ -212,7 +216,8 @@ runtime errors.
 
-The following tables show the operators that MUST be supported by a CESQL evaluator.
+The following tables show the operators that MUST be supported by a CESQL evaluator. When evaluating an operator,
+a CESQL engine MUST attempt to cast the operands to the correct types. If this type casting fails, a _Cast_ error will be returned, along with the zero value for the return type of the expression.
```
do we define the zero value for each type? While probably obvious we should say it's: zero, "" and `false`.

Cali0707 replied:
We do now :)

duglin commented on cesql/spec.md:
```diff
@@ -341,12 +346,11 @@ The following tables show the built-in functions that MUST be supported by a CES
 
-Operators MUST be evaluated in order, where the parenthesized expressions have the highest priority over all the other
-operators.
+Operators MUST be evaluated in left to right order, where all operators within a parenthized expression MUST be evaluated before continuing to the right of the parenthized expression.
```
parenthized -> parenthesized

Cali0707 replied:
yep I did :facepalm:

Cali0707 replied:
Fixed!

duglin commented:
Before we ship this - can we wrap it at 80 columns to match the other specs?

Cali0707 commented:
@duglin @jskeet would you be able to take another look when you have time? I _think_ I should have addressed all your feedback (minus wrapping at 80 characters), but I'm happy to make further fixes or address anything I missed.

Thanks again for all of your feedback so far!!

jskeet commented:
I'm unlikely to get time to review this again I'm afraid.

Cali0707 commented:
> I'm unlikely to get time to review this again I'm afraid.

No worries, thank you so much for taking the time the last few weeks to review this!

duglin commented on cesql/spec.md:
```diff
@@ -56,7 +56,10 @@ producer, or in an intermediary, and it can be implemented using any technology
-transparency][referential-transparency-wiki]. The output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_, and it might include an error.
+transparency][referential-transparency-wiki]. The primary output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_.
+The secondary output of a CESQL expression evaluation is a set of errors which occurred during evaluation. This set MAY be empty, indicating that no
+error occurred during execution of the expression. The value used by CESQL engines to represent an empty set of errors is out of the scope of this
```
Isn't the entire representation of, or even access to, errors out of scope of the spec, not just the empty set of errors?

duglin commented on cesql/spec.md:
```diff
@@ -197,8 +207,13 @@ Unless otherwise specified, every attribute and extension MUST be represented by
+is set to be the zero value for the type of the missing attribute. Rather, the subexpression with the missingAttribute returns the zero value of the return type of the subexpression.
+As an example, `1 / missingAttribute` does not raise a _MathError_ due to the division by zero, instead it returns `0` as that is the zero value for the return type of the subexpression.
+In cases where the return type of the subexpression cannot be determined by the CESQL engine, the CESQL engine MUST assume a return type of _Boolean_. In such cases, the return value would
+therefore be `false`.
```
geez I've been staring at this one for way too long....

duglin replied:
the wording is fine... it's the semantics that I keep thinking about.... :-)

Cali0707 replied:
Yeah, having thought about it more (and based on my comment in the other thread about missing attributes) I think moving back to missing attributes resulting in an error makes sense.

duglin commented on cesql/spec.md:
```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
+- _MissingFunctionError_: An error that occurs due to a call to a function that has not been registered with the CESQL engine
+- _FunctionEvaluationError_: An error that occurs during the evaluation of a function
+
+Whenever an operator or function encounters an error, it MUST return a return value as well as an error value or values. In cases where there is not an obvious return value for the expression,
```
See if this is easier to read:
...it MUST result in a "return value" as well as one or more "error values".

Up to you

duglin replied:
Is a missing attribute an error? Kind of feel like it should be and therefore should have an error type, no? I could be swayed into believing that it's special, but if so we may want to say that a missing attribute isn't an error in the spec so people don't think we overlooked it.

Also, I can't help but wonder if there should be some kind of "generic error" define for cases that you didn't think of above.

Cali0707 replied:
I keep going back to whether or not a missing attribute should be an error. I think with the introduction of zero values for each type along with the short circuiting of AND/OR, it makes sense to have missing attributes as an error again. A big part of the argument for not having an error was to try and understand an expression like:
```sql
missingAttribute AND true
```
as well as
```
true OR missingAttribute
```

But, with these new changes we have made I think that these can be reasoned about so including the error (with the zero type) makes more sense

duglin commented on cesql/spec.md:
```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
 
-The following tables show the operators that MUST be supported by a CESQL evaluator.
+The following tables show the operators that MUST be supported by a CESQL evaluator. When evaluating an operator,
+a CESQL engine MUST attempt to cast the operands to the correct types. A CESQL evaluator MUST stop evaluating an operator
```
s/correct/specified/

duglin commented on cesql/spec.md:
```diff
@@ -328,25 +363,31 @@ The following tables show the built-in functions that MUST be supported by a CES
-| `RIGHT(x, y): String x Integer -> String`                  | Returns a new string with the last `y` characters of `x` or returns `x` if `LENGTH(x) <= y`. Returns `x` if `y < 0` and raises an error.                                                                                   |
-| `SUBSTRING(x, pos): String x Integer x Integer -> String`  | Returns the substring of `x` starting from index `pos` (included) up to the end of `x`. Characters' index starts from `1`. If `pos` is negative, the beginning of the substring is `pos` characters from the end of the string. If `pos` is 0, then returns the empty string. Returns the empty string and raises an error if `pos > LENGTH(x) O ...
-| `SUBSTRING(x, pos, len): String x Integer x Integer -> String` | Returns the substring of `x` starting from index `pos` (included) of length `len`. Characters' index starts from `1`. If `pos` is negative, the beginning of the substring is `pos` characters from the end of the string. If `pos` is 0, then returns the empty string. If `len` is greater than the maximum substring starting at `pos`, t ...
+| `TRIM(x): String -> String`                                | Returns `x` with leading and trailing whitespaces trimmed. This does not remove any non-whitespace characters, such as control characters.                                                                                 |
```
I think we still need to define "whitespace characters" by pointing to some spec (or defining it here). Saying it's not control characters is misleading because `\t` is both a control character and whitespace

duglin commented on cesql/spec.md:
```diff
@@ -70,9 +73,8 @@ The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "S
-When used as a filter predicate, the expression output value is always cast to a boolean value.
-
-<!-- TODO -->
+When used as a filter predicate, the expression output value MUST be a _Boolean_. If the output value is not a _Boolean_ or any errors are returned,
```
add a comma after _Boolean_

duglin commented on cesql/spec.md:
```diff
@@ -56,7 +56,10 @@ producer, or in an intermediary, and it can be implemented using any technology
 type-checked CloudEvent instance. An expression MUST NOT mutate the value of the input CloudEvent instance, nor any of
 the other input values. The evaluation of an expression observes the concept of [referential
-transparency][referential-transparency-wiki]. The output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_, and it might include an error.
+transparency][referential-transparency-wiki]. The primary output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_.
```
Just curious, do you know why the reference to "referential transparency" is here? I'm just wondering what people should do with that info? If the sentence were removed would it hurt the spec or change the impls?

Cali0707 replied:
Oh as far as I know the point of this is that there are no side effects in a CESQL statement - every function/operator/subexpression can be replaced with the output of that function/operator/subexpression and you would get the same result

Cali0707 replied:
Thinking about this, I don't think it necessarily hurts the spec, but I find that it does give some clarity about the independence of parts of an expression. I'll try adding another sentence explaining the referential transparency

duglin commented on cesql/spec.md:
```diff
@@ -197,8 +207,13 @@ Unless otherwise specified, every attribute and extension MUST be represented by
 Through implicit type casting, the user can convert the addressed value instances to _Integer_ and
```
"... MUST be represented by the _String_ type as its initial type"...  I'm curious what this really means and why it's here. Does the user who is creating the expression need to do something with this info? Or is this for an implementation/SDK ? And if so, what do they do with it? If someone defines an extension called myint (an int), why would it matter if it starts out (internally) as an int or a string if it will be type cast as needed?

The spec says: The primary output of a CESQL expression evaluation is always a boolean, an integer or a string.

Does this mean that an expression of just `myint` MUST result in a string? And if so, it's kind of interesting that the spec doesn't define an int() function to allow people to covert it back to an int if that's what they want to result to be.

duglin replied:
I'm still not following. If I define an extension as an int, why must it be converted to a string ?
Especially if the cesql expression if `myint = 5` ?

duglin replied:
Let me ask this... what breaks if you just remove these sentences and this notion of "initial type"? Things will get converted automatically when needed. I'm just not seeing the point of (or the meaning of) the phrase "initial type".

Thinking about this more... is this exposing some implementation detail because the SDKs convert all attributes to strings initially? If so, then it really shouldn't be in the spec since the spec should focus on the cesql language syntax and resulting semantics, not how the code goes from "syntax" to "result".

Cali0707 replied:
How about "If the value of the attribute or extension is not one of the primitive CESQL types, it MUST be represented by the _String_ type"

duglin commented on cesql/spec.md:
```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
 
+Every CESQL engine MUST support the following error types:
+
+- _ParseError_: An error that occurs during parseing
```
s/parseing/parsing/

duglin commented on cesql/spec.md:
```diff
@@ -210,9 +225,24 @@ returned together with the evaluated value of the CESQL expression.
+as the error. In cases where the return type for the operator is ambiguous, the return type MUST be assumed to be
+_Boolean_ and the return value MUST be `false`.
 
 All the operators in the following tables are listed in precedence order.
```
Below, it defines "x = y" before "x != y". Given the sentence above, does this mean that:
```
a != b = c
```
must have "b = c" evaluated before the "a != ..." part?

The spec also says "Operators MUST be evaluated in left to right order", which I'd interpret as "a != b" would be evaluated first.

And I'm reading this wrong or are things a bit ambiguous?

Cali0707 replied:
@duglin I think that what actually happens is that subexpressions are evaluated left to right but that within that there is a precedence (coming from the antlr4 lexer for the language). Technically, what was originally written here is correct about the precedence of operations. Maybe we need to clarify that evaluation is left to right, but that the precedence of operators/functions takes priority over that L to R evaluation?

Cali0707 replied:
I've removed the text about operators being listed in precedence order, since it did not seem to match the actual precedence from the parser. Instead, I have added a table with the updated precedence.

duglin commented on cesql/spec.md:
```diff
@@ -229,38 +259,35 @@ Corresponds to the syntactic rule `unary-operation`:
+| `x / y: Integer x Integer -> Integer`   | Returns the result of dividing `x` by `y`, rounded towards `0` to obtain an integer. Returns `0` and a _MathError_ if `y = 0`      |
+| `x % y: Integer x Integer -> Integer`   | Returns the remainder of `x` divided by `y`, where the result has the same sign as `x`. Returns `0` and a _MathError_ if `y = 0`   |
+| `x + y: Integer x Integer -> Integer`   | Returns the sum of `x` and `y`                                                                                                     |
+| `x - y: Integer x Integer -> Integer`   | Returns the difference of `x` and `y`                                                                                              |
```
I'd prefer: Returns the value of `y` subtracted from `x`.   Or something like that.

"Returns the difference of x and y" is ambiguous as to which is subtracted from which. Technically, y-x is the diff too.

duglin commented on cesql/spec.md:
```diff
@@ -229,38 +259,35 @@ Corresponds to the syntactic rule `unary-operation`:
+| `x - y: Integer x Integer -> Integer`   | Returns the difference of `x` and `y`                                                                                              |
+| `x = y: String x String -> Boolean`     | Returns `true` if the values of `x` and `y` are equal                                                                              |
+| `x != y: String x String -> Boolean`    | Same as `NOT (x = y)`                                                                                                              |
+| `x <> y: String x String -> Boolean`    | Same as `NOT (x = y)`                                                                                                              |
```
Something for future consideration:  String > String, etc...

Cali0707 replied:
Yeah, we could add support for that in the future. For now, I'd prefer to keep it out of the spec and expand if there is a request

duglin commented on cesql/spec.md:
```diff
@@ -229,38 +259,35 @@ Corresponds to the syntactic rule `unary-operation`:
+| `x % y: Integer x Integer -> Integer`   | Returns the remainder of `x` divided by `y`, where the result has the same sign as `x`. Returns `0` and a _MathError_ if `y = 0`   |
+| `x + y: Integer x Integer -> Integer`   | Returns the sum of `x` and `y`                                                                                                     |
+| `x - y: Integer x Integer -> Integer`   | Returns the difference of `x` and `y`                                                                                              |
+| `x = y: String x String -> Boolean`     | Returns `true` if the values of `x` and `y` are equal                                                                              |
```
For easier reading, I'd put something like `(case sensitive)` on each of these lines so that people don't need to look outside of each row to get the full definition of the operator. Then you can delete the sentence you added on line 289.

duglin commented on cesql/spec.md:
```diff
@@ -278,6 +305,8 @@ For example, the pattern `_b*` will accept values `ab`, `abc`, `abcd1` but won't
 Both `%` and `_` can be escaped with `\`, in order to be matched literally. For example, the pattern `abc\%` will match
 `abc%` but won't match `abcd`.
 
+The pattern matching MUST be case sensitive.
```
As above, it might be nice to say it's "case sensitive" in each row rather than splitting the full definition between the table and a sentence 5 paragraphs away.

duglin commented on cesql/spec.md:
```diff
@@ -307,18 +337,23 @@ Because of implicit casting, no functions with the same name and same arity but
 
 An overload on a variadic function is allowed only if the number of initial fixed arguments is greater than the maximum arity for that particular function name. Only one variadic overload is allowed.
```
I think the wording here might need tweaking:
- "An overload on a variadic function"... reads to me as "You can create an overload of a function that's already defined with variadic args", which I don't think is the intent
- "Only one variadic overload is allowed" need to be scoped to a particular function name

Perhaps something like:
A function name MAY have at most one variadic overload definition and only if the number of initial fixed arguments is greater than the maximum arity of all other function definitions for that function name.

duglin commented on cesql/spec.md:
```diff
@@ -291,14 +320,15 @@ assumed valid, e.g. `EXISTS id` MUST always return `true`.
+| `x NOT IN (y1, y2, ...): Any x Any^n -> Boolean`       | Same as `NOT (x IN set)`                                                   |
 
 The matching is done using the same semantics of the equal `=` operator, but using `x` type as the target type for the implicit type casting.
 
```
If "n" can be zero, what does that mean for:
```
CONCAT(x1, x2, ...): String^n -> String
```
Is `CONCAT()` valid?  We may need to define the range of "n" on each use

Cali0707 replied:
For `CONCAT`, I think having no parameters is allowed, it would just return an empty string

duglin commented on cesql/spec.md:
```diff
@@ -328,25 +363,31 @@ The following tables show the built-in functions that MUST be supported by a CES
+
+As specified in 3.3, in the event of an error a function MUST still return a valid return value for it's defined return type.
+A CESQL engine MUST guarantee that all built-in functions comply with this. For user defined functions, if they return one or more errors
+and fail to providea valid return value for their return type the CESQL engine MUST return the zero value for the return type of the
```
s/providea/provide a/

duglin commented on cesql/spec.md:
```diff
@@ -355,26 +396,26 @@ left operand of the OR operation evalues to `true`, the right operand MUST NOT b
-   1. If it's not ambiguous, cast all the arguments to the corresponding parameter types.
-   1. If it's ambiguous, raise an error and the cast results are undefined.
+   1. Cast all the arguments to the corresponding parameter types.
 1. If the operator is n-ary with `n > 2`:
```
I'm sure there a reason, but why "n>2" and not "n>0" ?

And I guess on the previous one, why "n > 1" and not "n > 0" ?

Cali0707 replied:
For "n>2" instead of "n>0" this is because we already handled the "n=1" and "n=2" earlie with the unary and binary operator/function sections.

As for why we only handle binary operator resolution and not binary function resolution (the "n>1" vs "n>2"), I am not sure, this comes from before my time working on the spec. I'm inclined to believe there was a reason for it though

duglin commented on cesql/spec.md:
```diff
@@ -193,12 +205,20 @@ be used in the `IN` operator.
+is set to be the zero value for the type of the missing attribute. Rather, the subexpression with the missingAttribute returns the zero value of the return type of the subexpression.
+As an example, `1 / missingAttribute` does not raise a _MathError_ due to the division by zero, instead it returns `0, (missingAttributeError)` as that is the zero value for the return type of the subexpression.
+In cases where the return type of the subexpression cannot be determined by the CESQL engine, the CESQL engine MUST assume a return type of _Boolean_. In such cases, the return value would
+therefore be `false, (missingAttributeError)`.
```
These examples really help clarify things, even if I have to read it a few times to really let it sink in :-)

markpeek commented:
I appreciate the review and refinement that @jskeet @duglin @Cali0707 have been doing in this PR. Have the review and changes converged yet? If so, it might make sense for another week for the rest of the team to review before voting on the v1. Thoughts?

duglin commented on cesql/spec.md:
```diff
@@ -327,26 +362,47 @@ The following tables show the built-in functions that MUST be supported by a CES
+1. Function invocations
+1. Unary operators
+  1. NOT unary operator
+  1. `-` unary operator
```
I don't think you got the indenting you were hoping for, I see this:
<img width="440" alt="image" src="https://github.com/cloudevents/spec/assets/1944671/fc84a9d8-23c8-4288-9b88-8c78f5c640f4">

and this new direction is much clearer than what you had before.

Cali0707 commented:
> Have the review and changes converged yet? If so, it might make sense for another week for the rest of the team to review before voting on the v1. Thoughts?

To me it seems like we are converging, not sure if @duglin or @jskeet have other opinions. My thoughts would be that we should:
1. If everyone is okay with it in the meeting, merge these changes this week
2. Have another week for everyone to review the updated spec after these changes
3. Have the vote next Thursday

jskeet commented on cesql/spec.md:
```diff
@@ -56,7 +56,11 @@ producer, or in an intermediary, and it can be implemented using any technology
 type-checked CloudEvent instance. An expression MUST NOT mutate the value of the input CloudEvent instance, nor any of
 the other input values. The evaluation of an expression observes the concept of [referential
-transparency][referential-transparency-wiki]. The output of a CESQL expression evaluation is always a _boolean_, an _integer_ or a _string_, and it might include an error.
+transparency][referential-transparency-wiki]. This means that any part of an expression can be replaced with it's output
```
it's => its

jskeet commented on cesql/spec.md:
```diff
@@ -327,26 +359,47 @@ The following tables show the built-in functions that MUST be supported by a CES
+
+#### 3.5.3 Function Errors
+
+As specified in 3.3, in the event of an error a function MUST still return a valid return value for it's defined return type.
```
it's => its

duglin commented:
Approved on the 5/30 call.
Despite there being late changes to this PR (meaning, after Tuesday), we decided to merge it anyway so that people only need to look at the latest version of the spec to do their "v1.0" review. So, please look over the "main" branch version of the spec and add any comments to the "review issue": https://github.com/cloudevents/spec/issues/1282 

The current plans are to complete the final review by next week's call, and then (if the group approves), to start the official 1-week voting process.
